jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputInterceptorTest.java

```diff
@@ -0,0 +1,525 @@
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import javax.servlet.AsyncContext;
+import javax.servlet.ServletException;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.WriteListener;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.HttpTester;
+import org.eclipse.jetty.logging.StacklessLogging;
+import org.eclipse.jetty.server.handler.AbstractHandler;
+import org.eclipse.jetty.server.handler.HandlerCollection;
+import org.eclipse.jetty.util.Callback;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class HttpOutputInterceptorTest
+{
+    private static final Logger LOG = LoggerFactory.getLogger(HttpOutputInterceptorTest.class);
+
+    private Server _server;
+    private LocalConnector _localConnector;
+    private HandlerCollection _handlerCollection;
+    private StacklessLogging _stacklessLogging;
+
+    @BeforeEach
+    public void before() throws Exception
+    {
+        _server = new Server();
+        HttpConfiguration config = new HttpConfiguration();
+        config.setOutputBufferSize(1024);
+        config.setOutputAggregationSize(256);
+
+        _localConnector = new LocalConnector(_server, new HttpConnectionFactory(config));
+        _localConnector.setIdleTimeout(Duration.ofMinutes(1).toMillis());
+        _server.addConnector(_localConnector);
+        ServerConnector serverConnector = new ServerConnector(_server, new HttpConnectionFactory(config));
+        _server.addConnector(serverConnector);
+
+        _handlerCollection = new HandlerCollection();
+        _server.setHandler(_handlerCollection);
+        _stacklessLogging = new StacklessLogging(HttpChannel.class);
+    }
+
+    @AfterEach
+    public void after() throws Exception
+    {
+        _stacklessLogging.close();
+        _server.stop();
+    }
+
+    @Test
+    public void testInterceptorCallbackFailed() throws Exception
+    {
+        addInterceptor(nextInterceptor -> new HttpOutput.Interceptor()
```

Comment:
`addInterceptor()` seems unnecessary. Why don't you just add the interceptor from the `AbstractHandler` that you have in every test? In this way you also won't need `_handlerCollection`.

Something like:

```java
@Test
public void testInterceptorCallbackFailed() 
{
    start(new AbstractHandler()
    {
        @Override
        public void handle(...) 
        {
            HttpOutput httpOutput = baseRequest.getResponse().getHttpOutput();
            httpOutput.setInterceptor(...);
            ...
        }
    });
}

Reply:
Not all the interceptors added by the tests do the same thing, some fail the callback, some throw, some start a new thread and wait on a CountDownLatch before failing the callback.

Reply:
@lachlan-roberts you can do what you need with the interceptor on the second line of the handler, no?